src/actions/pkg.cr

```diff
@@ -1,45 +1,137 @@
-      next context
+
+    def clean_unused_packages(context : HTTP::Server::Context)
+      if result = get_prefix_or_default(from: context).clean_unused_packages(confirmation: false) { }
+        return {data: result} if result.any?
+        bug = InternalServerError.new context, "received empty set from Prefix#clean_unused_packages; please report this strange bug"
+        error = NoPkgsToClean.new context
+        error.cause = bug
+        raise error
+      end
+      raise NoPkgsToClean.new context
     end
-    raise Unauthorized.new context
-  end
-  # Query information about a given package
-  relative_get "/:id/query" do |context|
-    if context.current_user? && Config.has_access? context, Access::Read
-      # TODO: Query information about the given package
-      next context
+
+    # List built packages
+    relative_get nil do |context|
+      if context.current_user? && Config.has_access? context, Access::Read
+        # Build JSON directly to the HTTP response IO
+        build_json context.response do |json|
+          json.array do
+            # build an array of all the package names
+            get_prefix_or_default(from: context).each_pkg { |pkg| json.string pkg.package }
+          end
+        end
+        next context
+      end
+      raise Unauthorized.new context
     end
-    raise Unauthorized.new context
-  end
-  # Delete a given package
-  relative_delete "/:id/delete" do |context|
-    if context.current_user? && Config.has_access? context, Access::Delete
-      # TODO: Query information about the given package
-      next context
+    # Clean unused built packages
+    relative_delete "/clean" do |context|
+      if context.current_user? && Config.has_access? context, Access::Delete
+        clean_unused_packages(context).to_json context.response
+        context.response.flush
+        next context
+      end
+      raise Unauthorized.new context
     end
-    raise Unauthorized.new context
-  end
-  # Build a package, returning the ID of the built image, and perhaps a status
-  # message? We could also use server-side events or a websocket to provide the
-  # status of this action as it occurs over the API, rather than just returning
-  # a result on completion.
-  relative_post "/:id/build" do |context|
-    if context.current_user? && Config.has_access? context, Access::Create
-      # TODO: build the package based on the submitted configuration
+    # Query information about a given package
+    relative_get "/:id/query" do |context|
+      if context.current_user? && Config.has_access? context, Access::Read
+        package_name = URI.unescape context.params.url["id"]
+        version = context.params.query["version"]?.try { |v| URI.unescape v }
+        tag = context.params.query["tag"]?.try { |t| URI.unescape t }
+        # Iterate over the packages to find the relevant one.
+        selected_pkg = DPPM::Prefix::Pkg.create get_prefix_or_default(from: context),
```

Comment:
We don't want to create a new package, it already exists.
You can use `Prefix#new_pkg`.
I'll improve the documentation on this point.

Reply:
Maybe I can even delete `Pkg.create`, I'm working on that.

Reply:
Hm. Yeah, if the package is supposed to exist already, I don't think `new` or `create` are the best choices. Perhaps `load` or something like that?

Reply:
The problem is the `create` method. We should only use `Prefix` to create new package objects, existing or not.
Then, we can use `#exists?` or `#exists!` to check/ensure its presence.

Reply:
Ok excellent.